core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +94,224 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
                 .build();
         }
 
-        SdkHttpRequest signedRequest = doSign(request).build();
+        setParameters(request);
+
+        ContentChecksum contentChecksum = createChecksum(request);
+
+        SdkHttpRequest.Builder requestBuilder = sign(request, contentChecksum);
+
         return SyncSignedRequest.builder()
-            .request(signedRequest)
+            .request(requestBuilder.build())
             .payload(request.payload().orElse(null))
             .build();
     }
 
-    private SdkHttpRequest.Builder doSign(SyncSignRequest<? extends AwsCredentialsIdentity> request) {
-        String checksumHeaderName = request.property(CHECKSUM_HEADER_NAME);
-        ChecksumAlgorithm checksumAlgorithm = request.property(CHECKSUM_ALGORITHM);
-
-        if (StringUtils.isNotBlank(checksumHeaderName) && checksumAlgorithm == null) {
-            throw new IllegalArgumentException(
-                CHECKSUM_ALGORITHM + " cannot be null when " + CHECKSUM_HEADER_NAME + " is given!");
+    @Override
+    public AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request) {
+        if (CredentialUtils.isAnonymous(request.identity())) {
+            return AsyncSignedRequest.builder()
+                .request(request.request())
+                .payload(request.payload().orElse(null))
+                .build();
```

Comment:
I feel like this shouldn't need to be here. I think the caller should really be determining whether to sign a request or not. Thoughts?

Reply:
I have wondered about this too. But the caller is going to be identity type agnostic and this is AWSCredentials specific. Also, I'm not 100% sure at this point where _all_ Anonymous credentials can be set. Keeping this here seems fine.